### PR TITLE
Update build to use sbt 1.0

### LINKED ISSUE #1434: Upgrade the build to sbt 1.0
Work on building Akka with sbt 1.0 has begun (https://github.com/akka/akka/pull/23651) due to the synergy of nightly build we want to be able to also update Akka HTTP to build with sbt 1.0 to ensure nightly builds continue to work.

### PR DESCRIPTION
**Pull request by jonas**
Fixes #1434.

Includes the unidoc update changes from #925 which was reverted in #940 since it broke the  integrated build with akka/akka.

`sbt-jol` is removed based on https://github.com/akka/akka-http/issues/1434#issuecomment-331334611

### CODE REVIEW

File: akka-http2-support/build.sbt [hunk 1]
```diff
@@ -10,18 +10,19 @@ OSGi.http2Support
 fork in run in Test := true
 fork in Test := true
 
-connectInput in run in Test := true
+sbt.Keys.connectInput in run in Test := true
```
Comment (jonas) [inline akka-http2-support/build.sbt]: Due to
```
/opt/akka-http/akka-http2-support/build.sbt:13: error: reference to connectInput is ambiguous;
it is imported twice in the same scope by
import _root_.com.typesafe.sbt.MultiJvmPlugin.autoImport._
and import _root_.sbt.Keys._
connectInput in run in Test := true
^
```

**Comment by jrudolph** [comment]
Thanks a lot @jonas for getting the ball rolling here. Btw. it seems something is broken with MiMa as well:

```
[error] java.lang.NoSuchMethodError: com.typesafe.tools.mima.core.Settings.StringSetting(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)Lscala/tools/nsc/settings/MutableSettings$StringSetting;
[error] 	at com.typesafe.tools.mima.core.Settings.<init>(Settings.scala:14)
[error] 	at com.typesafe.tools.mima.core.Config$.setup(Config.scala:48)
[error] 	at com.typesafe.tools.mima.core.Config$.setup(Config.scala:45)
[error] 	at com.typesafe.tools.mima.plugin.SbtMima$.makeMima(SbtMima.scala:29)
[error] 	at com.typesafe.tools.mima.plugin.SbtMima$.runMima(SbtMima.scala:41)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11(MimaPlugin.scala:68)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11$adapted(MimaPlugin.scala:61)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$$$Lambda$4931/342874639.apply(Unknown Source)
[error] 	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$9(MimaPlugin.scala:61)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$9$adapted(MimaPlugin.scala:45)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$$$Lambda$3048/1455466014.apply(Unknown Source)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at scala.Function1$$Lambda$1009/748229733.apply(Unknown Source)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:42)
[error] 	at sbt.internal.util.$tilde$greater$$Lambda$1894/1532989443.apply(Unknown Source)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:64)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:257)
[error] 	at sbt.Execute$$Lambda$1911/270989935.apply(Unknown Source)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:266)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:257)
[error] 	at sbt.Execute$$Lambda$1902/948451889.apply(Unknown Source)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:167)
[error] 	at sbt.ConcurrentRestrictions$$anon$4$$Lambda$1909/430845669.apply(Unknown Source)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:32)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jonas** [comment]
Bumping mima to 0.1.18 fixes this.

**Comment by jonas** [comment]
Updating mima has also fixed the issue with running sbt-whitesource on Travis

 
```
[error] java.lang.NoSuchMethodError: sbt.compat$.ConfigTypeRef(Lsbt/librarymanagement/ConfigRef$;)Lsbt/librarymanagement/ConfigRef$;
[error] 	at sbtwhitesource.BaseAction.$anonfun$collectDependencyStructure$6(WhiteSource.scala:178)
...
[error] (akka-http-root/*:whitesourceCheckPolicies) java.lang.NoSuchMethodError: sbt.compat$.ConfigTypeRef(Lsbt/librarymanagement/ConfigRef$;)Lsbt/librarymanagement/ConfigRef$;
```
> https://travis-ci.org/akka/akka-http/builds/277802269#L808

**Comment by jonas** [comment]
Failure is: https://jenkins.akka.io:8498/job/pr-validator-akka-http/1564/testReport/akka.http.impl.engine.client/ConnectionPoolSpec/The_host_level_client_infrastructure_should_open_a_second_connection_if_the_first_one_is_loaded/

> Expected OnNext(_), yet no element signaled during 7500 milliseconds

**Comment by jonas** [comment]
PLS BUILD

File: akka-http2-support/build.sbt (hunk 1 above)
Comment (ktoso) [reply akka-http2-support/build.sbt]: :+1:

File: docs/src/test/scala/docs/http/scaladsl/HttpClientExampleSpec.scala [hunk 2]
```diff
@@ -148,10 +148,10 @@ class HttpClientExampleSpec extends WordSpec with Matchers with CompileOnlySpec
         val responseFuture: Future[HttpResponse] = dispatchRequest(HttpRequest(uri = "/"))
 
         responseFuture.andThen {
-          case Success(_) => println("request succeeded")
-          case Failure(_) => println("request failed")
+          case Success(_) ⇒ println("request succeeded")
```
Comment (ktoso) [inline docs/src/test/scala/docs/http/scaladsl/HttpClientExampleSpec.scala]: Hm... In Akka we usually do not want to show the fancy arrows in documentation; could we fix this before merging?

File: project/plugins.sbt [hunk 3]
```diff
@@ -6,21 +6,21 @@ resolvers += Resolver.sonatypeRepo("releases") // to more quickly obtain paradox
...
+addSbtPlugin("com.dwijnand" % "sbt-dynver" % "2.0.0")
+addSbtPlugin("com.typesafe.sbt" % "sbt-osgi" % "0.9.2")
+addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.4.1")
+addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings" % "2.0.0")
 addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.27")
-addSbtPlugin("pl.project13.sbt" % "sbt-jol" % "0.1.1")
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.8.2") // for advanced PR validation features
-addSbtPlugin("io.spray" % "sbt-boilerplate" % "0.6.0")
+//addSbtPlugin("pl.project13.sbt" % "sbt-jol" % "0.1.1")
+//addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.8.3-SNAPSHOT") // for advanced PR validation features
```
Comment (ktoso) [inline project/plugins.sbt]: please open a ticket about re-enabling this :)

File: project/plugins.sbt [hunk 4]
```diff
@@ -6,21 +6,21 @@ resolvers += Resolver.sonatypeRepo("releases") // to more quickly obtain paradox
...
+addSbtPlugin("org.scalariform" % "sbt-scalariform" % "1.8.0")
+addSbtPlugin("com.dwijnand" % "sbt-dynver" % "2.0.0")
+addSbtPlugin("com.typesafe.sbt" % "sbt-osgi" % "0.9.2")
+addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.4.1")
+addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings" % "2.0.0")
 addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.27")
-addSbtPlugin("pl.project13.sbt" % "sbt-jol" % "0.1.1")
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.8.2") // for advanced PR validation features
-addSbtPlugin("io.spray" % "sbt-boilerplate" % "0.6.0")
+//addSbtPlugin("pl.project13.sbt" % "sbt-jol" % "0.1.1")
```
Comment (ktoso) [inline project/plugins.sbt]: ok to just remove it

Comment (jonas) [reply project/plugins.sbt]: Done

File: project/plugins.sbt (hunk 3 above)
Comment (jonas) [reply project/plugins.sbt]: Sure, opened #1438

**Comment by jonas** [comment]
I haven't managed to get sbt-scalariform to work as before and am throwing the towel in the ring for now.

I suspect the following commit is part of the problem: https://github.com/sbt/sbt-scalariform/commit/42ca0f04b5e7280afcf8eec896580f141be85891, since it adds the project root directory as a source directory so that `docs` ends up being formatted with `RewriteArrowSymbol` enabled.

```
akka-http-root > compile:scalariformFormat::sourceDirectories
...
[info] akka-http-root/compile:scalariformFormat::sourceDirectories
[info] 	List(/opt/akka-http/src/main/scala-2.11, /opt/akka-http/src/main/scala, /opt/akka-http/src/main/java, /opt/akka-http)
akka-http-root > multi-jvm:scalariformFormat::sourceDirectories
[info] * /opt/akka-http/akka-http-tests/src/multi-jvm/scala-2.11
[info] * /opt/akka-http/akka-http-tests/src/multi-jvm/scala
[info] * /opt/akka-http/akka-http-tests/src/multi-jvm/java
[info] * /opt/akka-http
```

I'd appreciate another pair of eyes on this since I've tried a million things and am going crazy.

**Comment by jonas** [comment]
Thanks to @chbatey's work in https://github.com/akka/akka-http/pull/1489 the issues with the doc formatting is resolved!

**Comment by jrudolph** [comment]
I'll try publishing a better version of sbt-dependency-graph maybe even today.

**Comment by jonas** [comment]
That would be sweet @jrudolph. This PR needs to wait for akka/akka to also upgrade to sbt 1.0, right, to not break the nightly builds?

**Comment by jrudolph** [comment]
Just published another snapshot of sbt-dependency-graph with this version: `0.8.2+35-014e74c6`

**Comment by jrudolph** [comment]
It's in sonatype snapshots, so you might need an extra resolver.

**Comment by jonas** [comment]
Outstanding work @jrudolph! Updated sbt-dependency-graph to 0.9.0 and reenabled the PR validation code.

The PR is now ready. 🎉

**Comment by ktoso** [comment]
That's awesome, thanks!

File: build.sbt [hunk 5]
```diff
@@ -75,23 +75,26 @@ lazy val parsing = project("akka-parsing")
...
 lazy val http2Support = project("akka-http2-support")
-  .enablePlugins(JavaAgent)
+  .enablePlugins(JavaAgent, Unidoc)
   .disablePlugins(MimaPlugin) // experimental module still
   .settings(javaAgents += Dependencies.Compile.Test.alpnAgent)
   .dependsOn(httpCore, httpTestkit % "test", httpCore % "test->test")
   .addAkkaModuleDependency("akka-stream-testkit", "test")
 
 lazy val httpTestkit = project("akka-http-testkit")
+  .enablePlugins(Unidoc)
```
Comment (2m) [inline build.sbt]: Is this required when generating Javadoc with genjavadoc?

Comment (2m) [reply build.sbt]: I see now. When `akka.genjavadoc.enabled=true` then `Unidoc` requires `GenJavadocPlugin` and this enables it.

I think we should rename `Unidoc` plugin as it brings some confusion here. The sbt-unidoc plugin works at the root project. So enabling something called `Unidoc` on sub-projects is confusing. What about `BootstrapGenjavadoc`?

File: project/MultiNode.scala [hunk 6]
```diff
@@ -73,6 +73,14 @@ object MultiNode extends AutoPlugin {
         testResults.events ++ multiNodeResults.events,
         testResults.summaries ++ multiNodeResults.summaries)
     })
+
+  implicit class TestResultOps(val self: TestResult) extends AnyVal {
+    def id: Int = self match {
+      case TestResult.Passed ⇒ 0
+      case TestResult.Failed ⇒ 1
+      case TestResult.Error  ⇒ 2
```
Comment (2m) [inline project/MultiNode.scala]: Great investigative work! I will borrow this for the akka build conversion to sbt 1.x. :)

File: build.sbt (hunk 5 above)
Comment (2m) [reply build.sbt]: I have done it like so here: https://github.com/akka/akka/pull/23850/commits/8d597966866c8ea24c2340c47b1029f1426039bd

**Comment by jonas** [comment]
Thanks for the review and suggestion @2m.

**Comment by ktoso** [comment]
No reason to delay merging I think, thanks a ton!

**Pull request merged**